scheduler/kubernetes/kafka-k8s.md

```diff
@@ -0,0 +1,873 @@
+
+/opt/pwx/bin/pxctl v l | grep pvc-cc70447a-7c4b-11e7-a940-42010a8c0002
+733731201475618092      pvc-cc70447a-7c4b-11e7-a940-42010a8c0002        3 GiB   2       no      no              LOW             0       up - attached on 10.140.0.5
+
+
+kubectl exec -n kafka -it kafka-0 -- bash 
+
+bin/kafka-topics.sh --describe --zookeeper zk-headless.default.svc.cluster.local:2181 --topic px-kafka-topic
+
+Topic:px-kafka-topic    PartitionCount:3        ReplicationFactor:3     Configs:
+        Topic: px-kafka-topic   Partition: 0    Leader: 0       Replicas: 0,1,2 Isr: 0,1,2
+        Topic: px-kafka-topic   Partition: 1    Leader: 1       Replicas: 1,2,0 Isr: 1,2,0
+        Topic: px-kafka-topic   Partition: 2    Leader: 2       Replicas: 2,0,1 Isr: 2,0,1
+
+
+kubectl delete po/kafka-1 -n kafka
+pod "kafka-1" deleted
+
+kubectl get pods -n kafka -w
+NAME      READY     STATUS        RESTARTS   AGE
+kafka-0   1/1       Running           0         22h
+kafka-1   0/1       Terminating       0         22h
+kafka-2   1/1       Running           0         22h
+kafka-3   1/1       Running           0         18h
+kafka-1   0/1       Terminating       0         22h
+kafka-1   0/1       Terminating       0         22h
+kafka-1   0/1       Pending           0         0s
+kafka-1   0/1       Pending           0         0s
+kafka-1   0/1       Init:0/1          0         0s
+kafka-1   0/1       Init:0/1          0         2s
+kafka-1   0/1       PodInitializing   0         3s
+kafka-1   0/1       Running           0         5s
+kafka-1   1/1       Running           0         11s
+
+kubectl get pods -n kafka -o wide
+NAME      READY     STATUS    RESTARTS   AGE       IP           NODE
+kafka-0   1/1       Running   0          22h       10.0.160.3   k8s-2
+kafka-1   1/1       Running   0          1m        10.0.112.2   k8s-3
+kafka-2   1/1       Running   0          22h       10.0.64.4    k8s-1
+kafka-3   1/1       Running   0          18h       10.0.112.1   k8s-3
+
+bin/kafka-topics.sh --describe --zookeeper zk-headless.default.svc.cluster.local:2181 --topic px-kafka-topic
+
+Topic:px-kafka-topic    PartitionCount:3        ReplicationFactor:3     Configs:
+Topic: px-kafka-topic   Partition: 0    Leader: 0       Replicas: 0,1,2 Isr: 0,2,1
+Topic: px-kafka-topic   Partition: 1    Leader: 1       Replicas: 1,2,0 Isr: 2,0,1
+Topic: px-kafka-topic   Partition: 2    Leader: 2       Replicas: 2,0,1 Isr: 2,0,1
+
+```
+
+### Node Failover
+
+In the case of a statefulset if the node is unreachable, which could happen in either of two cases
+- The node is down for maintenance
+- There has been a network partition. 
+
+There is no way for kubernetes to know which of the case is it. Hence Kubernetes would not schedule the Statefulset and the pods running on those nodes would enter the ‘Terminating’ or ‘Unknown’ state after a timeout. For further information : [Statefulset Pod Deletion](https://kubernetes.io/docs/tasks/run-application/force-delete-stateful-set-pod/)
+
```

Comment:
Add a note that if the node reboots, the pod will terminate and get created somewhere else.